File: src/models/shared/rights/SiteRoles.php
<?php

namespace models\shared\rights;

use models\mapper\MapOf;

class SiteRoles extends RolesBase
{
    const SITE_MANAGER    = 'site_manager';    // highest possible level for site: various management responsibilities like creating projects
    const PROJECT_CREATOR = 'project_creator'; // permission to create a project
    const USER = 'user';
    const NONE = 'none';

    public static function init()
    {
        self::$_rights = array();

        // User
        $rights = array();
        $rights[] = Domain::USERS + Operation::EDIT_OWN;
        $rights[] = Domain::USERS + Operation::VIEW_OWN;
        $rights[] = Domain::PROJECTS + Operation::VIEW_OWN;
        // Should users be able to delete their own user accounts? Probably,
        // but not via the listview -- so we should NOT grant DELETE_OWN here.
        //$rights[] = Domain::USERS + Operation::DELETE_OWN;
        self::$_rights[self::USER] = $rights;

        // Project Creator (User plus ability to create projects)
        $rights = self::$_rights[self::USER];
        $rights[] = Domain::PROJECTS + Operation::CREATE;
        $rights[] = Domain::PROJECTS + Operation::ARCHIVE_OWN;
        self::$_rights[self::PROJECT_CREATOR] = $rights;

        // Site Manager (all rights on projects)
        $rights = self::$_rights[self::USER];
        self::grantAllOnDomain($rights, Domain::PROJECTS);
        self::$_rights[self::SITE_MANAGER] = $rights;

    }

    private static $_rights;

    /**
     *
     * @param MapOf $roleMap
     * @param Website $website
     * @return array
     */
    public static function getRightsArray($roleMap, $website)
    {
        if ($roleMap->offsetExists($website->domain)) {
            return self::_getRightsArray(self::$_rights, $roleMap[$website->domain]);
        }

        return array();
    }

    /**
     *
     * @param  MapOf      $roleMap
     * @param  int        $right
     * @throws \Exception
     * @internal param Website $website
     * @return bool
     */
    public static function hasRight($roleMap, $right)
    {
        global $WEBSITE;
        if ($roleMap->offsetExists($WEBSITE->domain)) {
            return self::_hasRight(self::$_rights, $roleMap[$WEBSITE->domain], $right);
        }

        return false;
    }

}
SiteRoles::init();

File: src/SIL.XForge.WebApi.Server/Models/SiteRoles.cs
using System.Collections.Generic;

namespace SIL.XForge.WebApi.Server.Models
{
    public class SiteRoles : RolesBase
    {
        public const string SiteManager = "site_manager";
        public const string ProjectCreator = "project_creator";
        public const string User = "user";
        public const string None = "none";

        public static SiteRoles Instance { get; } = new SiteRoles();

        private SiteRoles()
        {
            var userRights = new HashSet<Right>
            {
                new Right(Domain.Users, Operation.EditOwn),
                new Right(Domain.Users, Operation.ViewOwn),
                new Right(Domain.Projects, Operation.ViewOwn)
            };
            Rights[User] = userRights;

            var projectCreatorRights = new HashSet<Right>(userRights)
            {
                new Right(Domain.Projects, Operation.Create)
            };
            Rights[ProjectCreator] = projectCreatorRights;

            var siteManagerRights = new HashSet<Right>(projectCreatorRights);
            siteManagerRights.UnionWith(AllRights(Domain.Projects));
            Rights[SiteManager] = siteManagerRights;
        }

        public bool HasRight(UserEntity user, string site, Right right)
        {
            if (user.SiteRole.TryGetValue(site, out string siteRole))
                return HasRight(siteRole, right);
            return false;
        }

        public bool HasRight(string siteRole, Right right)
        {
            if (siteRole == null)
                return false;
            return Rights[siteRole].Contains(right);
        }
    }
}
